### ISSUE TITLE
Reduce with array input defaults the return type to an array of arrays

### ISSUE THREAD

**Issue by PrimalZed** [ReactiveX/rxjs#2519]
**RxJS version:** 5.3.0

**Code to reproduce:**
```
Observable.of([0, 1, 2, 3],[4, 5, 6, 7])
.reduce((acc, value) => acc.concat(value), [])
.map(numbers => numbers.map(n => n++))
.subscribe(value => console.log(JSON.stringify(value)));
```

**Expected behavior:** Can take an input type T[] and return type T[] without specifying typing.

**Actual behavior:** On input type T[], defaults return type T[][].

**Additional information:**
This was introduced with the fix to #2382 
My current workaround is to simply specify the types, e.g. `.reduce<number[], number[]>(...)`

**Comment by aluanhaddad** [comment]
@PrimalZed The issue you are experiencing comes from the type of `of`, not from `reduce`. The reduce issue, where all seed parameters were optional, with `Array<ObservableElementType>` taking precedence was probably hiding the underlying issue.
For reference, here is the error I receive.
![image](https://cloud.githubusercontent.com/assets/5034656/24635121/94c76098-189f-11e7-8907-d0e4516257ec.png)

**Comment by PrimalZed** [comment]
@aluanhaddad the type from `of` was intentional to be `Observable<number[]>`.  The `reduce` is to combine all the arrays returned by the observable into a single array, with `Array.prototype.concat`.

That error is the problem I'm experiencing, yes.  The return type of `reduce` is `Observable<number[][]>` rather than the desired `Observable<number[]>`.

Again, it's easily resolved by specifying some typing somewhere, but it seems like it should be able to pick it up.  Maybe a type definition of `reduce<T>(this: Observable<T[]>, accumulator: (acc: T[], value: T[], index: number) => T[], seed: T[]): Observable<T[]>` should be added?  Or is this use case just too fringe to deserve a type definition to support it?

**Comment by david-driscoll** [comment]
The overload was added so that common case of reduce to an array of `T` would be easily supported without explicit casting.  It looks like in this case however it falls over.

```typescript
// without the overload
Observable.of(0, 1, 2, 3)
  .reduce((acc, value) => acc.concat(value), []); // [] would be {}[] and would require a cast to T[] to work correctly.
```

It appears that reducing the precedence of the array variant we can get things to work.  See the [playground here](http://www.typescriptlang.org/play/#src=interface%20ObservableStatic%20%7B%0D%0A%20%20%20%20of%3CT%3E(...args%3A%20T%5B%5D)%3A%20Observable%3CT%3E%3B%0D%0A%7D%0D%0Ainterface%20Observable%3CT%3E%20%7B%0D%0A%20%20%20%20map%3CT%2C%20R%3E(this%3A%20Observable%3CT%3E%2C%20project%3A%20(value%3A%20T%2C%20index%3A%20number)%20%3D%3E%20R%2C%20thisArg%3F%3A%20any)%3A%20Observable%3CR%3E%3B%0D%0A%20%20%20%20subscribe(fn%3A%20Function)%3B%0D%0A%20%20%20%20reduce%3CT%3E(this%3A%20Observable%3CT%3E%2C%20accumulator%3A%20(acc%3A%20T%2C%20value%3A%20T%2C%20index%3A%20number)%20%3D%3E%20T%2C%20seed%3F%3A%20T)%3A%20Observable%3CT%3E%3B%0D%0A%20%20%20%20reduce%3CT%3E(this%3A%20Observable%3CT%3E%2C%20accumulator%3A%20(acc%3A%20T%5B%5D%2C%20value%3A%20T%2C%20index%3A%20number)%20%3D%3E%20T%5B%5D%2C%20seed%3A%20T%5B%5D)%3A%20Observable%3CT%5B%5D%3E%3B%0D%0A%20%20%20%20reduce%3CT%2C%20R%3E(this%3A%20Observable%3CT%3E%2C%20accumulator%3A%20(acc%3A%20R%2C%20value%3A%20T%2C%20index%3A%20number)%20%3D%3E%20R%2C%20seed%3A%20R)%3A%20Observable%3CR%3E%3B%0D%0A%7D%0D%0A%0D%0Alet%20Observable%3A%20ObservableStatic%3B%0D%0A%0D%0AObservable.of(%5B0%2C%201%2C%202%2C%203%5D%2C%20%5B4%2C%205%2C%206%2C%207%5D)%0D%0A%20%20%20%20.reduce((acc%2C%20value)%20%3D%3E%20acc.concat(value)%2C%20%5B%5D)%0D%0A%20%20%20%20.map(numbers%20%3D%3E%20numbers.map(n%20%3D%3E%20n%2B%2B))%0D%0A%20%20%20%20%2F%2F%20.subscribe(value%20%3D%3E%20console.log(JSON.stringify(value)))%3B%0D%0A%0D%0AObservable.of(0%2C%201%2C%202%2C%203%2C%204%2C%205%2C%206%2C%207)%0D%0A%20%20%20%20.reduce((acc%2C%20value)%20%3D%3E%20%7B%20acc.push(value)%3B%20return%20acc%3B%20%7D%2C%20%5B%5D)%0D%0A%20%20%20%20.map(numbers%20%3D%3E%20numbers.map(n%20%3D%3E%20n%2B%2B))%0D%0A%20%20%20%20%2F%2F%20.subscribe(value%20%3D%3E%20console.log(JSON.stringify(value)))%3B) or the code below.  All that was done was that `reduce<T>(this: Observable<T>, accumulator: (acc: T[], value: T, index: number) => T[], seed: T[]): Observable<T[]>;` was moved down one line.

```typescript
interface ObservableStatic {
    of<T>(...args: T[]): Observable<T>;
}
interface Observable<T> {
    map<T, R>(this: Observable<T>, project: (value: T, index: number) => R, thisArg?: any): Observable<R>;
    subscribe(fn: Function);
    reduce<T>(this: Observable<T>, accumulator: (acc: T, value: T, index: number) => T, seed?: T): Observable<T>;
    reduce<T>(this: Observable<T>, accumulator: (acc: T[], value: T, index: number) => T[], seed: T[]): Observable<T[]>;
    reduce<T, R>(this: Observable<T>, accumulator: (acc: R, value: T, index: number) => R, seed: R): Observable<R>;
}

let Observable: ObservableStatic;

Observable.of([0, 1, 2, 3], [4, 5, 6, 7])
    .reduce((acc, value) => acc.concat(value), [])
    .map(numbers => numbers.map(n => n++))
    // .subscribe(value => console.log(JSON.stringify(value)));

Observable.of(0, 1, 2, 3, 4, 5, 6, 7)
    .reduce((acc, value) => { acc.push(value); return acc; }, [])
    .map(numbers => numbers.map(n => n++))
    // .subscribe(value => console.log(JSON.stringify(value)));
```

**Comment by PrimalZed** [comment]
Thanks @david-driscoll, I was wondering if the type definitions worked as a "pick first matching" sort of thing.  So all we have to do is change the order of the type definitions.

**Comment by aluanhaddad** [comment]
@david-driscoll thanks that explains it. Otherwise you need to write `[] as number[]` to make your example typecheck under `--noImplicitAny`

@PrimalZed Essentially it picks the first matching overload yes.

**Comment by kwonoj** [comment]
closed via #2523.
